OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -434,20 +484,41 @@ public void QueueOrder(Order order)
 			orders.Enqueue(order);
 		}
 
+		// Moved the Check for if a unit is buildable to where it was selected instead of built. ~~StormWing0
 		ActorInfo ChooseRandomUnitToBuild(ProductionQueue queue)
 		{
 			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+
+			if (buildableThings == null || !buildableThings.Any())
 				return null;
 
-			var unit = buildableThings.Random(Random);
+			// ~~StormWing0 Added new checks for if a unit can be built or trained and moved some older checks into it
+			var availibleUnits = buildableThings.Where(actorUnit =>
```

Comment:
avalaibleUnits

Reply:
oops didn't see that. >.<

Reply:
Oh I remember what I was trying to do. The reason I went that route was I wanted the AI to check what it could build for units first before trying to build the unit since it kept locking up the AI attempting to build the same unit over and over again.  So I figured I'd make it so it'd only pick from what it could build instead of picking from everything than checking if it could make it.

I think I can narrow down the lines of code to cut out a ton of that junk so it doesn't look so odd.